pkg/build/builder/daemonless.go

```diff
@@ -0,0 +1,293 @@
+package builder
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/containers/image/pkg/docker/config"
+	"github.com/containers/image/pkg/sysregistriesv2"
+	"github.com/containers/image/transports/alltransports"
+	"github.com/containers/image/types"
+	"github.com/containers/storage"
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	buildapiv1 "github.com/openshift/api/build/v1"
+	"github.com/pkg/errors"
+	"github.com/projectatomic/buildah"
+	"github.com/projectatomic/buildah/imagebuildah"
+	"github.com/projectatomic/buildah/util"
+)
+
+var (
+	// TODO: run unprivileged https://github.com/openshift/origin/issues/662
+	daemonlessStoreOptions = storage.DefaultStoreOptions
+)
+
+func checkForDaemonlessImage(imageName string) error {
+	glog.V(0).Infof("Checking if %q has already been pulled....", imageName)
+
+	store, err := storage.GetStore(daemonlessStoreOptions)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if _, err := store.Shutdown(true); err != nil {
+			glog.V(0).Infof("Error shutting down storage: %v", err)
+		}
+	}()
+
+	var systemContext *types.SystemContext
+	_, img, err := util.FindImage(store, "", systemContext, imageName)
+	if err != nil {
+		switch errors.Cause(err) {
+		case storage.ErrImageUnknown, docker.ErrNoSuchImage:
+			glog.V(0).Infof("Local copy of %q is not present.", imageName)
+			return docker.ErrNoSuchImage
+		}
+		return err
+	}
+	if img == nil {
+		return docker.ErrNoSuchImage
+	}
+
+	glog.V(0).Infof("Local copy of %q is present.", imageName)
+	return err
+}
+
+func pullDaemonlessImage(imageName string, authConfig docker.AuthConfiguration) error {
+	glog.V(0).Infof("Asked to pull fresh copy of %q.", imageName)
+
+	_, err := alltransports.ParseImageName("docker://" + imageName)
+	if err != nil {
+		return err
+	}
+
+	glog.V(0).Infof("Deferring pull of %q.", imageName)
+	return nil
+}
+
+func buildDaemonlessImage(dir string, optimization buildapiv1.ImageOptimizationPolicy, opts *docker.BuildImageOptions) error {
+	glog.V(0).Infof("Building...")
+
+	args := make(map[string]string)
+	for _, ev := range opts.BuildArgs {
+		args[ev.Name] = ev.Value
+	}
+
+	pullPolicy := buildah.PullIfMissing
+	if opts.Pull {
+		glog.V(0).Infof("Forcing fresh pull of base image.")
+		pullPolicy = buildah.PullAlways
+	}
+
+	squash := false
+	layers := false
+	switch optimization {
+	case buildapiv1.ImageOptimizationDaemonless:
+		// layers: false, squash: false
+	case buildapiv1.ImageOptimizationDaemonlessWithLayers:
+		layers = true
+	case buildapiv1.ImageOptimizationDaemonlessSquashed:
+		squash = true
+	default:
+		return fmt.Errorf("internal error: image optimization policy %q not fully implemented", string(optimization))
+	}
+
+	var systemContext types.SystemContext
+	if credsDir, ok := os.LookupEnv("PULL_DOCKERCFG_PATH"); ok {
+		systemContext.AuthFilePath = filepath.Join(credsDir, "config.json")
```

Comment:
I actually thought this might be unnecessary, since `pkg/build/builder/cmd/dockercfg` seems handle parsing secrets in multiple formats and hands us structures with the values in them that, when we pass them to the image library, are used in preference to any files that we tell it about.

The image library thinks of .dockercfg as a legacy format, so I didn't see the file parsing logic in origin as something that we'd be dropping until upstream dropped it as a supported secrets type.